src/mavlink-router/mainloop.cpp

```diff
@@ -37,69 +36,48 @@
 #define TIMEOUT_LOG_SHUTDOWN_US     5000000ULL // number of microseconds we wait until we give up trying to stop log streaming
                                         // after a shutdown of mavlink router was requested
 
-static std::atomic<bool> should_exit {false};
-
-Mainloop Mainloop::_instance{};
-bool Mainloop::_initialized = false;
-
-static void exit_signal_handler(int signum)
-{
-    Mainloop::request_exit();
-}
-
-static void setup_signal_handlers()
-{
-    struct sigaction sa = { };
-
-    sa.sa_flags = SA_NOCLDSTOP;
-    sa.sa_handler = exit_signal_handler;
-    sigaction(SIGTERM, &sa, nullptr);
-    sigaction(SIGINT, &sa, nullptr);
-
-    sa.sa_handler = SIG_IGN;
-    sigaction(SIGPIPE, &sa, nullptr);
-}
-
-Mainloop &Mainloop::init()
-{
-    assert(_initialized == false);
-
-    _initialized = true;
-
-    return _instance;
-}
-
 static const char* pipe_path = "/tmp/mavlink_router_pipe";
 
-void Mainloop::request_exit()
-{
-    should_exit.store(true, std::memory_order_relaxed);
-}
+Mainloop* Mainloop::instance = nullptr;
 
-int Mainloop::open()
+Mainloop::Mainloop()
 {
-    if (epollfd != -1)
-        return -EBUSY;
+    if (instance) {
+        throw std::logic_error("Only one mainloop instance must exist at a given time");
+    }
```

Comment:
I guess somewhere you need a:

```suggestion
    }
    instance = this;
```

I didn't see it anywhere else

Reply:
yes good catch that is wrong indeed -- I guess it should go to the very end, when all setup is done